src/contractor/graph_contractor.cpp

```diff
@@ -568,28 +547,216 @@ bool GraphContractor::IsNodeIndependent(const std::vector<float> &priorities,
+                                  node_data.priorities[node] = EvaluateNodePriority(
+                                      SimulateNodeContraction(data, graph, node, node_data.weights),
+                                      node_data.depths[node]);
+                              }
+                          });
+        log << " ok.";
+    }
+
+    auto number_of_core_nodes = std::max<std::size_t>(0, (1 - core_factor) * number_of_nodes);
+    auto number_of_nodes_to_contract = remaining_nodes.size() - number_of_core_nodes;
+    util::Log() << "preprocessing " << number_of_nodes_to_contract << " ("
+                << (number_of_nodes_to_contract / (float)number_of_nodes * 100.) << "%) nodes...";
+
+    util::UnbufferedLog log;
+    util::Percent p(log, remaining_nodes.size());
+
+    const util::XORFastHash<> hash;
+
+    unsigned current_level = 0;
+    std::size_t next_renumbering = number_of_nodes * 0.35;
+    while (remaining_nodes.size() > number_of_core_nodes)
+    {
+        if (remaining_nodes.size() < next_renumbering)
+        {
+            RenumberData(remaining_nodes, new_to_old_node_id, node_data, graph);
+            log << "[renumbered]";
+            // only one renumbering for now
+            next_renumbering = 0;
+        }
+
+        tbb::parallel_for(
+            tbb::blocked_range<NodeID>(0, remaining_nodes.size(), IndependentGrainSize),
+            [&](const auto &range) {
+                ContractorThreadData *data = thread_data_list.GetThreadData();
+                // determine independent node set
+                for (auto i = range.begin(), end = range.end(); i != end; ++i)
+                {
+                    const NodeID node = remaining_nodes[i].id;
+                    remaining_nodes[i].is_independent =
+                        IsNodeIndependent(hash, node_data.priorities, graph, data, node);
+                }
+            });
+
+        // sort all remaining nodes to the beginning of the sequence
+        const auto begin_independent_nodes =
+            stable_partition(remaining_nodes.begin(),
+                             remaining_nodes.end(),
+                             [](RemainingNodeData node_data) { return !node_data.is_independent; });
+        auto begin_independent_nodes_idx =
+            std::distance(remaining_nodes.begin(), begin_independent_nodes);
+        auto end_independent_nodes_idx = remaining_nodes.size();
+
+        // contract independent nodes
+        tbb::parallel_for(
+            tbb::blocked_range<NodeID>(
+                begin_independent_nodes_idx, end_independent_nodes_idx, ContractGrainSize),
+            [&](const auto &range) {
+                ContractorThreadData *data = thread_data_list.GetThreadData();
+                for (auto position = range.begin(), end = range.end(); position != end; ++position)
+                {
+                    const NodeID node = remaining_nodes[position].id;
+                    node_data.is_core[node] = false;
```

Comment:
here is a ub due to N4687 26.2.2 /2

Reply:
Gah yeah forgot vector<bool> is not thread save. Okay yeah that explains a lot of weird bugs I have been seeing.